dcp-0009/dcp-0009.mediawiki

```diff
@@ -0,0 +1,279 @@
+<pre>
+DCP: 0009
+Title: Automatic Ticket Revocations
+Author: Ryan Staudt <mail@ryanstaudt.com>
+Status: Defined
+Created: 2021-08-25
+License: CC0-1.0
+License-Code: ISC
+</pre>
+
+==Abstract==
+
+This proposes modifications to Decred ticket revocation transactions and block
+acceptance criteria in order to support automatic ticket revocations.
+
+The main changes are:
+
+* Requiring that revocation transactions are version <code>2</code> transactions
+* Requiring that revocation transaction inputs do not have signature scripts
+* Requiring the revocation transaction fee to be zero
+* Updating the ticket redeemer output amounts calculation for revocations
+* Skipping script validation for revocation transactions
+* Requiring blocks to contain revocation transactions for all tickets that will become missed or expired as of that block
+
+==Motivation==
+
+Two primary goals motivate the proposed changes:
+
+* Improve the Decred stakeholder user experience by removing the requirement for stakeholders to manually revoke missed and expired tickets
+* Enable the recovery of funds for users who lost their redeem script for the legacy VSP system (before the release of vspd, which removed the need for the redeem script)
+
+==Specification==
+
+===Revocation Transactions===
+
+This proposal defines the following rule updates for revocation transactions:
+
+* The revocation transaction version MUST be <code>2</code>
```

Comment:
I don't have a strong opinion on this, and I'm fine with changing it to `3`, but I went with `2` for the following reasons:
- Revocation transactions are currently version `1`, so it seems natural to bump them to version `2`
- There is no benefit of bumping it higher without the explicit versions consensus change (since revocation transactions of any higher version will be accepted currently)
- The updated consensus rules are enforcing that the version MUST be `2`, so even with the explicit versions consensus change, version `3` revocation transactions will be rejected when this agenda is active
- Even if we made it `3` to align with the max version allowed by the explicit versions consensus change, when another future agenda is approved that bumps the overall max transaction version but does impact revocations, you'll still have the "wiggle room" where you could have a higher version revocation transaction allowed by the overall max transaction version rule, but still rejected by the revocation specific consensus rules

Reply:
I'm fine with going with `2` since I don't have a super strong preference here and there is no technical issue with it.  It's a fair point that now that the revocation version is going to enforced explicitly by the transaction type itself (contrary to the current rules), there is a bit of separation versus the overall explicit consensus version change for transactions as a whole.  I tend to prefer consistency with versions so it's less cognitive overhead to try and remember exactly which version each individual type of thing is and requires less special casing, but given the code is already written and tested with version `2`, that probably isn't a good enough justification for changing it.